Provide a way to reset attribute changes

Draggha opened the pull request:
Hello there!

Previously the reset function overwrote the original attribute state with the current one making it a save() without a request and making all future calls to isDirty() invalid.

This should fix #21.

Draggha commented:
Oooh nice, now I see that you also use `reset()` after `save()`. Now that is a bit inconvenient since [your documentation clearly states that your `reset()` reverts all changes to the attributes](https://github.com/graphiti-api/graphiti-api.github.io/blob/master/js/writes/dirty-tracking.md).

wadetandy commented:
@Draggha When you first submitted this issue, I thought you must be doing something wrong, as I was sure that `reset()` behaved as the documentation says, and always had, but sure enough it seems that in the time between when that method was written and documented, the meaning changed in our head 👎 .  We definitely need functionality like this, and a PR that added a method to accomplish it would be great. Changing `reset()` to handle this is probably a bad idea since there's no telling who's using it for the currently defined behavior, so I'm open to suggestions.  `revert()` is probably ambiguous with reset but perhaps `restoreOriginal()` or something?

richmolj commented:
Yeah, I think this is a documentation issue - thanks for raising. The docs are incorrect - `#reset()` means "reset the dirty tracking" not "reset to the original state". We want `#reset()` to continue working with the current behavior, but need a separate method like `#clean()` or `#restore()` or `#rollback()` that will restore original state.

In the meantime, I've updated the documentation to be correct. @Draggha  would you be willing to submit a PR for this?

Draggha commented:
Hello there!

I updated my branch to add a `rollback` function which just resets changes to the attributes. relationship changes would need a deeper understanding of the code base, so I didn't add it on top.
I also updated the title to reflect the current state of this feature.

Thank you all for the insights and the guidance! :)

- J

P.S.: You added `yarn.lock` to your `.gitignore` file, so you won't have version lockdowns (since you still use version ranges in your package versions). Was this intended?

richmolj commented:
This looks ready for merge to me. I think the method will eventually need to `rollback(true)` to reset relationships as well, but seems OK in the meantime. @wadetandy any objections?

> P.S.: You added yarn.lock to your .gitignore file, so you won't have version lockdowns (since you still use version ranges in your package versions). Was this intended?

Yep - this way we test with the latest version `package.json` says is allowed, and catch issues users might experience earlier.

richmolj commented:
@wadetandy bump

wadetandy commented:
sorry about that, must have missed the last updates in my inbox.
